### PR TITLE
Adding ClientSideRequestStatistics to gateway calls and making endtime nullable

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

CosmosClientSideRequestStatistics end time was configured to be the start time. This causes confusion when there was never any calls updating the endtime. Making it optional makes it clear the endtime was never set.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs [hunk 1]
```diff
@@ -324,6 +324,13 @@ private static async Task<Stream> BufferContentIfAvailableAsync(HttpResponseMess
                     DateTime sendTimeUtc = DateTime.UtcNow;
                     Guid localGuid = Guid.NewGuid();  // For correlating HttpRequest and HttpResponse Traces
 
+                    IClientSideRequestStatistics clientSideRequestStatistics = request.RequestContext.ClientRequestStatistics;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs]: Why to sprinkle these across callers? Why not the Context do initialization?

File: Microsoft.Azure.Cosmos/src/Resource/Settings/CosmosClientSideRequestStatistics.cs [hunk 2]
```diff
@@ -51,7 +51,12 @@ public TimeSpan RequestLatency
         {
             get
             {
-                return this.requestEndTime - this.requestStartTime;
+                if (this.requestEndTime.HasValue)
+                {
+                    return this.requestEndTime.Value - this.requestStartTime;
+                }
+
+                return TimeSpan.Zero;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/CosmosClientSideRequestStatistics.cs]: ZERO is not representative and might be mis-leading as well. How about making it TimeSpan.MAX or something like that?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDiagnosticsTests.cs [hunk 3]
```diff
@@ -180,6 +179,7 @@ public static void VerifyPointDiagnostics(CosmosDiagnostics diagnostics)
             // Session token only expected on success
             if (statusCode >= 200 && statusCode < 300)
             {
+                Assert.IsNotNull(jObject["requestEndTimeUtc"].ToString());
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDiagnosticsTests.cs]: Why conditional?

Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDiagnosticsTests.cs]: On exception scenarios the end time is never updated.

File: Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs]: DocumentServiceRequest is inside the direct package so we have no way to create it as part of the Context. There is also no central factory for creating DocumentServiceRequest. The `ClientRetryPolicy` populates it for most scenarios, but not all scenarios use that retry policy like the caches.

**Pull request merged**